src/SourceBuild/content/eng/tools/tasks/Microsoft.DotNet.UnifiedBuild.Tasks/JoinVerticals.cs

```diff
@@ -0,0 +1,317 @@
+            if (verticalName == MainVertical)
+            {
+                continue;
+            }
+
+            addedPackageIds = AddMissingElements(packageElements, verticalManifest, _packageElementName);
+            addedBlobIds = AddMissingElements(blobElements, verticalManifest, _blobElementName);
+
+            if (addedPackageIds.Count > 0 || addedBlobIds.Count > 0)
+            {
+                downloadTasks.Add(
+                    DownloadArtifactFiles(
+                        BuildId,
+                        $"{verticalName}{_artifactNameSuffix}",
+                        addedPackageIds,
+                        addedBlobIds,
+                        clientThrottle));
+            }
+        }
+
+        await Task.WhenAll(downloadTasks);
+
+        // Create MergedManifest.xml
+        // taking the attributes from the main manifest
+        XElement mainManifestRoot = verticalManifests.First().Root
+            ?? throw new ArgumentException("The root element of the vertical manifest is null.");
+        mainManifestRoot.Attribute(_verticalNameAttribute)!.Remove();
+
+        string manifestOutputPath = Path.Combine(OutputFolder, "MergedManifest.xml");
+        XDocument mergedManifest = new(new XElement(
+            mainManifestRoot.Name,
+            mainManifestRoot.Attributes(),
+            packageElements.Values.Select(v => v.Element).OrderBy(elem => elem.Attribute(_idAttribute)?.Value),
+            blobElements.Values.Select(v => v.Element).OrderBy(elem => elem.Attribute(_idAttribute)?.Value)));
+
+        File.WriteAllText(manifestOutputPath, mergedManifest.ToString());
+
+        Log.LogMessage(MessageImportance.High, $"### Duplicate items found in the following verticals: ###");
+
+        foreach (var item in duplicatedItems)
+        {
+            Log.LogMessage(MessageImportance.High, $"Item: {item.Key} -- Produced by: {string.Join(", ", item.Value)}");
+        }
+    }
+
+    /// <summary>
+    /// Downloads specified packages and symbols from a specific build artifact and stores them in an output folder, 
+    /// either flat or with the same relative path as in the artifact.
+    /// </summary>
+    private async Task DownloadArtifactFiles(
+        string buildId,
+        string artifactName,
+        List<string> packageFileNames,
+        List<string> assetFileNames,
+        SemaphoreSlim clientThrottle)
+    {
+        string packagesOutputPath = Path.Combine(OutputFolder, _packagesFolderName);
+        string assetsOutputPath = Path.Combine(OutputFolder, _assetsFolderName);
+
+        using AzureDevOpsClient azureDevOpsClient = new(AzureDevOpsToken, AzureDevOpsBaseUri, AzureDevOpsProject, Log);
+
+        try
+        {
+            await clientThrottle.WaitAsync();
+
+            ArtifactFiles filesInformation = await azureDevOpsClient.GetArtifactFilesInformation(buildId, artifactName, _retryCount);
+
+            foreach (var package in packageFileNames)
```

Comment:
Can you put a note here that additional parallelism is available to Task.WhenAll the DownloadFile* tasks (and move the client throttle down into there). You say it got down to 5 mins so most likely there isn't a ton of additional win here, but something to keep in mind.

Reply:
I went ahead and implemented that, let me know if it looks right